src/vcpkg/binarycaching.cpp

```diff
@@ -1020,34 +1020,243 @@ namespace
+                }
+            }
+        }
+
+        void precheck(View<const InstallPlanAction*> actions, Span<CacheAvailability> cache_status) const override
         {
+            auto abis = azcopy_list();
+            if (abis.empty())
+            {
+                // If the command failed, we assume that the cache is unavailable.
+                std::fill(cache_status.begin(), cache_status.end(), CacheAvailability::unavailable);
+                return;
+            }
+
+            for (size_t idx = 0; idx < actions.size(); ++idx)
+            {
+                auto&& action = *actions[idx];
+                const auto& abi = action.package_abi().value_or_exit(VCPKG_LINE_INFO);
+                cache_status[idx] =
+                    Util::contains(abis, abi) ? CacheAvailability::available : CacheAvailability::unavailable;
+            }
         }
 
-        std::string make_object_path(const std::string& prefix, const std::string& abi) const override
+        LocalizedString restored_message(size_t count,
+                                         std::chrono::high_resolution_clock::duration elapsed) const override
         {
-            return Strings::replace_all(prefix, "{sha}", abi);
+            return msg::format(
+                msgRestoredPackagesFromAzureStorage, msg::count = count, msg::elapsed = ElapsedTime(elapsed));
         }
+
+        Path m_buildtrees;
+        AzCopyUrl m_url;
+        Path m_tool;
     };
-    struct AzCopyStoragePushProvider : ObjectStoragePushProvider
+    struct AzCopyStoragePushProvider : IWriteBinaryProvider
     {
-        AzCopyStoragePushProvider(std::vector<std::string>&& prefixes,
-                                  const std::shared_ptr<const IObjectStorageTool>& tool)
-            : ObjectStoragePushProvider(std::move(prefixes), tool)
+        AzCopyStoragePushProvider(std::vector<AzCopyUrl>&& containers, const Path& tool)
+            : m_containers(std::move(containers)), m_tool(tool)
         {
         }
 
-        std::string make_object_path(const std::string& prefix, const std::string& abi) const override
+        vcpkg::ExpectedL<Unit> upload_file(StringView url, const Path& archive) const
         {
-            return Strings::replace_all(prefix, "{sha}", abi);
+            auto upload_cmd = Command{m_tool}
+                                  .string_arg("copy")
+                                  .string_arg("--from-to")
+                                  .string_arg("LocalBlob")
+                                  .string_arg(archive)
+                                  .string_arg(url);
+
+            return flatten(cmd_execute_and_capture_output(upload_cmd), Tools::AZCOPY);
+        }
+
+        size_t push_success(const BinaryPackageWriteInfo& request, MessageSink& msg_sink) override
+        {
+            if (!request.zip_path) return 0;
+            const auto& zip_path = *request.zip_path.get();
```

Comment:
```suggestion
            const auto& zip_path = request.zip_path.value_or_exit(VCPKG_LINE_INFO);
```

This thing asks for zips. If we get here and there aren't zips something bad has happened and we need to die.